src/app/core/services/kafka.service.ts

```diff
@@ -156,73 +140,75 @@ export class KafkaService {
-    let schemaVersions
-    switch (specs.name) {
-      case KAFKA_COMPLETION_LOG:
-        if (this.schemas[specs.name]) {
-          schemaVersions = this.schemas[specs.name]
-          break
-        }
-      default:
-        schemaVersions = this.util
-          .getLatestKafkaSchemaVersions(specs)
-          .catch(error => {
-            console.log(error)
-            return Promise.resolve()
-          })
-        this.schemas[specs.name] = schemaVersions
-    }
-    return Promise.all([schemaVersions]).then(data => {
-      schemaVersions = data[0]
-      const avroKey = AvroSchema.parse(
-        JSON.parse(schemaVersions[0]['schema']),
-        {
-          wrapUnions: true
-        }
-      )
-      const avroVal = AvroSchema.parse(
-        JSON.parse(schemaVersions[1]['schema']),
-        {
-          wrapUnions: true
-        }
-      )
-      const kafkaObject = specs.kafkaObject
-      const bufferKey = avroKey.clone(kafkaObject.key, { wrapUnions: true })
-      const bufferVal = avroVal.clone(kafkaObject.value, { wrapUnions: true })
-      const payload = {
-        key: bufferKey,
-        value: bufferVal
+    return this.util.getKafkaTopic(specs).then(topic => {
+      let schemaVersions
+      switch (specs.name) {
+        case KAFKA_COMPLETION_LOG:
+          if (this.schemas[specs.name]) {
+            schemaVersions = this.schemas[specs.name]
+            break
+          }
+        default:
+          schemaVersions = this.util
+            .getLatestKafkaSchemaVersions(topic)
+            .catch(error => {
+              console.log(error)
+              return Promise.resolve()
+            })
+          this.schemas[specs.name] = schemaVersions
       }
-      const schemaId = new KafkaRest.AvroSchema(
-        JSON.parse(schemaVersions[0]['schema'])
-      )
-      const schemaInfo = new KafkaRest.AvroSchema(
-        JSON.parse(schemaVersions[1]['schema'])
-      )
-      return this.sendToKafka(
-        specs,
-        schemaId,
-        schemaInfo,
-        payload,
-        kafkaConnInstance
-      ).catch(error => {
-        console.error(
-          'Could not initiate kafka connection ' + JSON.stringify(error)
+      return Promise.all([schemaVersions]).then(data => {
+        schemaVersions = data[0]
+        const avroKey = AvroSchema.parse(
+          JSON.parse(schemaVersions[0]['schema']),
+          {
+            wrapUnions: true
+          }
+        )
+        const avroVal = AvroSchema.parse(
+          JSON.parse(schemaVersions[1]['schema']),
+          {
+            wrapUnions: true
+          }
+        )
+        const kafkaObject = specs.kafkaObject
+        const bufferKey = avroKey.clone(kafkaObject.key, { wrapUnions: true })
+        const bufferVal = avroVal.clone(kafkaObject.value, { wrapUnions: true })
+        const payload = {
+          key: bufferKey,
+          value: bufferVal
+        }
+        const schemaId = new KafkaRest.AvroSchema(
```

Comment:
This part of code is really confusing. How can we fetch schemaKey and schemaValue then schemaId and schemaInfo from the same data. Maybe adding some documentation and proper variable naming would help.

Reply:
Yes I have also already started refactoring in #249. I've divided the Kafka service into 2 services: kafka service and schema service (https://github.com/RADAR-base/RADAR-Questionnaire/pull/249/files#diff-cdb09557472fa72feae59cc1f4e80091R1) so that the schema conversion is separate and more readable. I haven't added documentation so I will do that.

This should be merged after `enable-reminders`.